src/server/pps/server/pipeline_controller.go

```diff
@@ -295,22 +296,22 @@ func (pc *pipelineController) step(timestamp time.Time) (isDelete bool, retErr e
 	// the pipeline indefinitely)
 	rc, restart, err := pc.getRC(ctx, pi)
 	if restart {
-		if err := pc.restartPipeline(ctx, pi, rc, "could not get RC."); err != nil {
+		if err := pc.restartPipeline(ctx, pi, rc); err != nil {
 			return false, err
 		}
-		return false, pc.setPipelineState(ctx, pi.SpecCommit, pps.PipelineState_PIPELINE_RESTARTING, "could not get RC.")
+		return false, errors.EnsureStack(pc.psDriver.SetState(ctx, pi.SpecCommit, pps.PipelineState_PIPELINE_RESTARTING, "could not get RC."))
 	}
 	if err != nil && !errors.Is(err, errRCNotFound) {
 		return false, err
 	}
-	targetState, sideEffects, err := evaluate(pi, rc)
+	targetState, sideEffects, reason, err := evaluate(pi, rc)
 	if err != nil {
 		return false, err
 	}
 	errCount := 0
 	err = backoff.RetryNotify(func() error {
 		// Create/Modify/Delete pipeline resources as needed per new state
-		return pc.apply(ctx, pi, rc, targetState, sideEffects)
+		return pc.apply(ctx, pi, rc, targetState, sideEffects, reason)
```

Comment:
hmm, feels like we narrowly avoided a bug before in the case of an external RC deletion while RUNNING? `getRC` would return a nil `rc` but `restart` would be `false`, `evaluate` would say we need to restart, `apply` would error on the restart because of how that function worked. It was okay because we errored out of the backoff, too, and would just print the failure at the top level. Still makes me a little nervous

Reply:
right (yikes)